### PR TITLE
[WIP] Lazy iterators

### LINKED ISSUE #7303: Lazily initialize iterators within series and shards
### Feature Request

Opening a feature request kicks off a discussion.
Requests may be closed if we're not actively planning to work on them.

**Proposal:** Lazily initialize iterators within each series and shard to lower the memory usage of the query engine when processing high cardinality data and make queries with limits faster.

**Current behavior:** Every shard iterator is initialized at the same time and there is a separate iterator for every series within every shard. With high cardinality data, this uses up a lot of memory as a baseline for even initializing the query. We also do not keep the returned series separate from each other when returning data from the shard which causes limit queries to return at least one data point from every shard within the time period rather than only reading the points it needs.

**Desired behavior:** We should only initialize the iterators we need. That means we need to organize them by the returned series and only initialize them when they are needed. They will also need to be sorted ahead of time by the shard start and end times.

**Use case:** This is required for LIMIT to run quickly when given an unbounded time and also for preventing OOM errors when a query has high cardinality.

### PR DESCRIPTION
**Pull request by jsternberg**
Fixes #7303.

### CODE REVIEW

File: coordinator/shard_mapping.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+					shardics[i] = shard
+				}
+				inner[i] = influxql.LazyIteratorCreator(shardics)
+			}
+			outer = append(outer, influxql.LazyIteratorCreator(inner))
+		}
+		ics[i] = influxql.IteratorCreators(outer)
+	}
+	return influxql.NewLazyIterator(ics, opt)
+}
```
Comment (benbjohnson) [inline coordinator/shard_mapping.go]: Can this function be broken up into some smaller functions?

File: coordinator/shard_mapping.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	SeriesKey   []byte
+	TagSet      *influxql.TagSet
+	StartTime   time.Time
+}
+
+func (si *SeriesInfo) CreateIterator(opt influxql.IteratorOptions) (influxql.Iterator, error) {
+	return si.Shard.CreateSeriesIterator(si.Measurement, si.TagSet, opt)
+}
+
+type byTime []*SeriesInfo
```
Comment (benbjohnson) [inline coordinator/shard_mapping.go]: Can you rename this to `seriesInfoByTime`?

File: influxql/iterator.gen.go [hunk 3]
```diff
@@ -561,6 +561,115 @@ func (itr *floatLimitIterator) Next() (*FloatPoint, error) {
...
+// newFloatLazyIterator creates a new instance of floatLazyIterator.
+func newFloatLazyIterator(input FloatIterator, ics IteratorCreators, opt IteratorOptions) *floatLazyIterator {
+	return &floatLazyIterator{
+		input: input,
+		ics:   ics,
+		opt:   opt,
+	}
+}
+
+// Next
```
Comment (benbjohnson) [inline influxql/iterator.gen.go]: ?

File: coordinator/shard_mapping.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+			for _, sh := range group {
+				shards = append(shards, sh.Shard)
+			}
+		}
+		return influxql.NewLazyIterator(shards, opt)
+	}
+
+	var mu sync.Mutex
+	parallelism := runtime.GOMAXPROCS(0)
+	ch := make(chan struct{}, parallelism)
```
Comment (jwilder) [inline coordinator/shard_mapping.go]: You could use `pkg/limiter` here.

File: coordinator/shard_mapping.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+)
+
+type Source struct {
+	Database        string
+	RetentionPolicy string
+}
+
+type ShardInfo struct {
+	Shard        Shard
+	Measurements []string
```
Comment (jwilder) [inline coordinator/shard_mapping.go]: Can you add some docs about what these fields are?  Is `Measurements` the subset of the measurements in `Shard` that need to be queried against?

File: coordinator/statement_executor.go [hunk 6]
```diff
@@ -596,14 +587,58 @@ func (e *StatementExecutor) createIterators(stmt *influxql.SelectStatement, ctx
...
+			source := Source{
+				Database:        s.Database,
+				RetentionPolicy: s.RetentionPolicy,
+			}
+
+			shards, ok := a[source]
+			if !ok {
+				groups, err := e.MetaClient.ShardGroupsByTimeRange(s.Database, s.RetentionPolicy, opt.MinTime, opt.MaxTime)
+				if err != nil {
+					return nil, err
```
Comment (jwilder) [inline coordinator/statement_executor.go]: Should this be a `continue`?

File: tsdb/store.go [hunk 7]
```diff
@@ -773,7 +773,45 @@ func (s *Store) deleteSeries(database string, seriesKeys []string, min, max int6
...
 // ExpandSources expands sources against all local shards.
 func (s *Store) ExpandSources(sources influxql.Sources) (influxql.Sources, error) {
-	return s.IteratorCreators().ExpandSources(sources)
+	m := make(map[string]influxql.Source)
+
+	if err := func() error {
+		s.mu.RLock()
+		defer s.mu.RUnlock()
+
+		for _, sh := range s.shards {
```
Comment (jwilder) [inline tsdb/store.go]: This could use `walkShards` to do this concurrently.

File: influxql/iterator.gen.go (hunk 3 above)
Comment (jsternberg) [inline influxql/iterator.gen.go]: I just never finished writing the comment here.

**Comment by jsternberg** [comment]
While the code will still need to be cleaned up with us possibly renaming of some of the structs and seeing where we can speed it up with parallelization, I think the overall structure is correct. @benbjohnson can you take another look at this? @desa my tools for benchmarking are limited since we don't seem to have too many benchmark tests in `server_test.go`, but can you try running this through our normal stress tests to ensure I didn't impact performance?

**Comment by desa** [comment]
@jsternberg yeah, what specifically should I keep an eye on?

**Comment by jsternberg** [comment]
@desa I'm more concerned about query return speed, but also try to keep track of memory usage if you can.

**Comment by desa** [comment]
## Results

This was ran against my laptop. If you'd like I can run them on a more isolated environment and monitor memory and what not.

```
Lazy Query: 'SELECT count(n) FROM ctr'
max: 130.662389ms
avg: 108.673896ms
min: 103.895517ms

Master Query: 'SELECT count(n) FROM ctr'
max: 53.127637ms
avg: 31.602136ms
min: 24.349265ms
```

```
Lazy Query: 'SELECT max(n) FROM ctr'
max: 122.211096ms
avg: 113.858843ms
min: 109.67186ms

Master Query: 'SELECT max(n) FROM ctr'
max: 37.011238ms
avg: 32.121948ms
min: 27.692981ms
```

```
Lazy Query: 'SELECT count(n) FROM ctr where time > now() - 1m'
max: 128.244992ms
avg: 59.896173ms
min: 49.038675ms

Master Query: 'SELECT count(n) FROM ctr where time > now() - 1m'
max: 185.018828ms
avg: 45.844356ms
min: 19.587185ms
```

```
Lazy Query: 'SELECT max(n) FROM ctr where time > now() - 1m'
max: 158.765362ms
avg: 70.531538ms
min: 53.773342ms

Master Query: 'SELECT max(n) FROM ctr where time > now() - 1m'
max: 28.301979ms
avg: 24.217428ms
min: 20.991712ms
```

```
Lazy Query: 'SELECT count(n) FROM ctr group by *'
max: 237.561556ms
avg: 233.305712ms
min: 228.495463ms

Master Query: 'SELECT count(n) FROM ctr group by *'
max: 106.078982ms
avg: 98.91626ms
min: 93.403709ms
```

```
Lazy Query: 'SELECT max(n) FROM ctr group by *'
max: 241.456428ms
avg: 238.674588ms
min: 234.88566ms

Master Query: 'SELECT max(n) FROM ctr group by *'
max: 103.909965ms
avg: 100.098236ms
min: 96.929756ms
```

```
Lazy Query: 'SELECT top(n,10) FROM ctr group by *'
max: 655.135983ms
avg: 631.452003ms
min: 608.059305ms

Master Query: 'SELECT top(n,10) FROM ctr group by *'
max: 1.210739386s
avg: 1.176479096s
min: 1.149928495s
```

```
Lazy Query: 'SELECT count(n) FROM ctr where time > now() - 1m group by time(10s), *'
max: 296.444553ms
avg: 176.4545ms
min: 140.906963ms

Master Query: 'SELECT count(n) FROM ctr where time > now() - 1m group by time(10s), *'
max: 339.99122ms
avg: 224.392393ms
min: 182.735858ms
```

```
Lazy Query: 'SELECT max(n) FROM ctr where time > now() - 1m group by time(10s), *'
max: 292.438719ms
avg: 188.165604ms
min: 151.526146ms

Master Query: 'SELECT max(n) FROM ctr where time > now() - 1m group by time(10s), *'
max: 332.474196ms
avg: 241.989004ms
min: 187.667ms
```

```
Lazy Query: 'SELECT top(n,10) FROM ctr where time > now() - 1m group by time(10s), *'
max: 683.436762ms
avg: 568.339036ms
min: 501.529034ms

Master Query: 'SELECT top(n,10) FROM ctr where time > now() - 1m group by time(10s), *'
max: 1.442596211s
avg: 1.22785927s
min: 1.06010002s
```


## Repro Steps
#### 1. Load 10M points
Run `influx-stress insert -n 10000000 ctr,host=server,location=us-west` and wait for it to finish.

#### 2. Run the script to get latencies
Run the script found in this [gist](https://gist.github.com/desa/6cc4e093afc06cc72450b7ca8f2233d2).

When the script says `Start stress...`, run the following stress test
```
influx-stress insert ctr,host=server,location=us-west
```
and wait for the test to finish.

**Comment by jsternberg** [comment]
Those results... confuse me. So it seems like it performs much slower when you have to return one aggregated point, but it returns moderately faster when returning with a `GROUP BY time(...)`. I'm also confused why `SELECT count(n) FROM ctr where time > now() - 1m'` is faster, but `SELECT max(n) FROM ctr group by *` is slower.

**Comment by desa** [comment]
Yeah I'm with you. They confused me as well.

**Comment by jwilder** [comment]
Performance on my local testing for a basic count of 10M points across 100k series is 2x slower for some reason.  Need to figure out what's going on there before we can merge this.

**Comment by jsternberg** [comment]
We'll revisit this, but this PR isn't good since it makes the current performance slower. The change to avoid instantiating key cursors/iterators for shards that do not have a series works better for low cardinality data and makes this PR not as urgent.

**Pull request closed without merging**